Support filtering out retrieved EC2 dynamic inventory instances in ec2.ini

michalgasek opened the pull request:
This allows filtering out EC2 instances based on various different filters including tags. As requested in #7480 it supports logical "OR" instead of "AND" on the provided list of filters.

Multiple API calls are made if the list of filters in ec2.ini config contains more than 1 different filter key.
See examples in ec2.ini for more information.

michalgasek commented:
I suppose a bit of clarification might be good here.

Amazon EC2 API action "DescribeInstances" supports only "AND" operation for multiple filters keys in a single API request. This behaviour is probably best explained in the 2nd example request in the EC2 API doc ([link](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/ApiReference-query-DescribeInstances.html#ApiReference-query-DescribeInstances-Example-Request-2)).
"ORing" in a single API call is however done in a special case: when provided filter contains only 1 key and multiple values for the same key are asked for.

This pull request will "OR" on every filter provided in `instance_filters` list in `ec2.ini` configuration file.

To put it simply: 1 unique filter key = 1 extra API request.
Table below shows the number of requests needed on examples to achieve this "ORing".

| `instance_filters` value in ec2.ini | EC2 API reqs # |
| --- | --: |
| `tag:role=webservers` | 1 |
| `instance-type=t1.micro` | 1 |
| `tag:role=webservers,tag:role=dbservers` | 1 |
| `tag:role=webservers,instance-type=t1.micro` | 2 |
| `tag:role=webservers,instance-type=t1.micro,instance-type=m1.small` | 2 |
| `tag:role=webservers,instance-type=t1.micro,architecture=i386` | 3 |
| `tag:role=webservers,tag:env=stage,instance-type=m1.small` | 3 |

Hope I did not confuse much :)

Since this is my first contribution to Ansible project - please let me know if anything is not right!

tgerla commented:
Excellent!! This is better patch than mine in 7480 since it supports multiple or'd filters. I will try to test this next week.

Commit:
Add support for filtering EC2 instances in dynamic inventory

This allows filtering out EC2 instances based on various different
filters including tags. As requested in 7480 it supports logical "OR"
instead of "AND" on the provided list of filters.

michalgasek commented:
Managed to break my branch on the way while merging/rebasing/resetting... Above commit fixes it so everything should be fine now.

mpdehaan commented:
Up-prioritizing this as it will be popular, thanks!

abadger commented:
Hey @tgerla , was this something you wanted to test?  If so I'll leave it be for a little while to give you a chance to look at it.  Thanks!

jimi-c commented:
# Merged

Hi!

This has been merged in.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
